plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/OpenshiftUIModel.java

```diff
@@ -0,0 +1,189 @@
+import org.jboss.tools.openshift.common.core.connection.IConnectionsRegistryListener;
+import org.jboss.tools.openshift.core.connection.IOpenShiftConnection;
+
+import com.openshift.restclient.model.IResource;
+
+public class OpenshiftUIModel implements IOpenshiftUIElement<IOpenshiftUIElement<?>> {
+	private Map<IOpenShiftConnection, ConnectionWrapper> connections = new HashMap<>();
+	private List<IElementListener> listeners = new ArrayList<IElementListener>();
+
+	private IConnectionsRegistryListener listener;
+	
+	public OpenshiftUIModel() {
+		this(ConnectionsRegistrySingleton.getInstance());
+	}
+
+	public OpenshiftUIModel(ConnectionsRegistry registry) {
+		listener = new IConnectionsRegistryListener() {
+
+			@Override
+			public void connectionRemoved(IConnection connection) {
+				synchronized (connections) {
+					connections.remove(connection);
+				}
+				fireChanged();
+			}
+
+			@Override
+			public void connectionChanged(IConnection c, String property, Object oldValue, Object newValue) {
+				ConnectionWrapper connection = connections.get(c);
+				if (connection == null) {
+					return;
+				}
+				connection.connectionChanged(property, oldValue, newValue);
+			}
+
+			@Override
+			public void connectionAdded(IConnection connection) {
+				if (!(connection instanceof IOpenShiftConnection)) {
+					return;
+				}
+				synchronized (connections) {
+					if (connections.containsKey(connection)) {
+						return;
+					}
+					connections.put((IOpenShiftConnection) connection,
+							new ConnectionWrapper(OpenshiftUIModel.this, (IOpenShiftConnection) connection));
+				}
+				fireChanged();
+			}
+		};
+		Collection<IConnection> allConnections = registry.getAll();
+		synchronized (connections) {
+			for (IConnection connection : allConnections) {
+				if (connection instanceof IOpenShiftConnection) {
+					connections.put((IOpenShiftConnection) connection,
+							new ConnectionWrapper(OpenshiftUIModel.this, (IOpenShiftConnection) connection));
+				}
+			}
+		}
+		registry.addListener(listener);
+	}
+	public static boolean isOlder(IResource oldResource, IResource newResource) {
+		try {
+			int oldVersion = Integer.valueOf(oldResource.getResourceVersion());
+			int newVersion = Integer.valueOf(newResource.getResourceVersion());
+			return oldVersion < newVersion;
+		} catch (NumberFormatException e) {
+			// treat this as an update
+			return true;
+		}
+	}
+
+	@Override
+	public IOpenshiftUIElement<?> getParent() {
+		return null;
+	}
+
+	@Override
+	public void fireChanged(IOpenshiftUIElement<?> source) {
+		if (Display.getCurrent() != null) {
+			dispatchChange(source);
+		} else {
+			Display.getDefault().asyncExec(new Runnable() {
+
+				@Override
+				public void run() {
+					dispatchChange(source);
+				}
+			});
+		}
+	}
+
+	protected void dispatchChange(IOpenshiftUIElement<?> source) {
+		Collection<IElementListener> copy = new ArrayList<>();
+		synchronized (listeners) {

```

Comment:
If we want the collection to remain the same across multiple operations (see  removeListener()), then it makes a difference.